src/main.c

```diff
@@ -142,10 +147,160 @@ static bool is_valid_stream(char *s)
+	const char* xdg_state_home = getenv("XDG_STATE_HOME");
+	const char* home_path = getenv("HOME");
+
+	if (!home_path & !xdg_state_home)
+	exit(EXIT_FAILURE);
+
+	const char* local_state = "/.local/state";
+	const char* dir_name = "/nudoku/";
+
+	char* dir_path = NULL;
+
+	if (xdg_state_home)
+	{
+	// user does have $XDG_STATE_HOME
+	size_t len_xdg_state_home = strlen(xdg_state_home);
+	size_t len_dir_name = strlen(dir_name);
+	dir_path = malloc(len_xdg_state_home + len_dir_name + 1);
+
+	strcpy(dir_path, xdg_state_home);
+	strcat(dir_path, dir_name);
+	}
+	else
+	{
+	// fallback to $HOME/.local/state
+	size_t len_home_path = strlen(home_path);
+	size_t len_local_state = strlen(local_state);
+	size_t len_dir_name = strlen(dir_name);
+	dir_path = malloc(len_home_path + len_local_state + len_dir_name + 1);
+
+	strcpy(dir_path, home_path);
+	strcat(dir_path, local_state);
+	strcat(dir_path, dir_name);
+	}
+
+	struct stat st = {0};
+	if (stat(dir_path, &st) == -1)
+	{
+	mkdir(dir_path, 0700);
+	}
+
+	const char* file_name = STATE_FILE_NAME;
+	size_t len_file = strlen(file_name);
+	size_t len_dir_path = strlen(dir_path);
+
+	char* file_path = malloc(len_dir_path + len_file + 1);
+	strcpy(file_path, dir_path);
+	strcat(file_path, file_name);
+
+	free(dir_path);
+	return file_path;
+}
+
+bool is_saved_game_valid(void)
+{
+	char* file_path = get_saved_file_path();
+	FILE* fp = fopen(file_path, "r");
+
+	if (fseek(fp, STREAM_LENGTH, SEEK_SET) != 0)
+	{
+	fclose(fp);
+	free(file_path);
+	return false;
+	}
+
+	char test_board[STREAM_LENGTH];
+	int c;
+	int i = 0;
+
+	while (i < STREAM_LENGTH)
+	{
+	c = fgetc(fp);
+	if (c == EOF) break;
+	test_board[i++] = (char)c;
+	}
+
+	bool valid = is_valid_stream(test_board);
+
+	fclose(fp);
+	free(file_path);
+
+	return valid;
+}
+
+void save_stream(char user_board[], char plain_board[], int n)
+{
+	char* file_path = get_saved_file_path();
+	FILE* fp;
+	fp = fopen(file_path, "w");
+	if ( fp == NULL )
+		exit(EXIT_FAILURE);
+	for ( int i=0; i<n; ++i )
+	{
+		fprintf(fp, "%c", user_board[i]);
+	}
+	for ( int i=0; i<n; ++i )
+	{
+		fprintf(fp, "%c", plain_board[i]);
+	}
+	fclose(fp);
+	free(file_path);
+}
+
+void get_saved_game_state(char user_board[], char plain_board[])
+{
+	FILE* fp;
+	char* file_path = get_saved_file_path();
+	fp = fopen(file_path, "r");
+
+	int c;
+	int i = 0;
+	while ( i < STREAM_LENGTH )
+	{
+		c = fgetc(fp);
+		user_board[i++] = (char)c;
+	}
+
+	i = 0;
+	while ( ( c = fgetc(fp)) != EOF )
+	{
+		plain_board[i++] = (char)c;
+	}
+
+	fclose(fp);
+	unlink(file_path); // remove file after resuming the game
+	free(file_path);
+
+	// set difficulty level
+	int count = 0;
+	const char *tmp = plain_board;
+	while((tmp = strchr(tmp, '.')) != NULL)
+	{
+	count++;
+	tmp++;
+	}
+
+	if (count == 50)
+		g_level = D_HARD;
+	else if (count == 40)
+		g_level = D_NORMAL;
+	else
+		g_level = D_EASY;
+}
+
+static void print_resume_error(void)
+{
+	printf(_("Game save is missing or corrupted, try starting new game\n"));
```

Comment:
This function just consists of a oneliner, that is even used only one singel time. So I don't think we need a function for it.